### ISSUE TITLE
Can't start_tls with raise_exceptions==True

### ISSUE THREAD

**Issue by ace6807** [cannatag/ldap3#454]
When I create a Connection with raise_exceptions=True, start_tls() returns false. As soon as set raise_exceptions=False, start_tls() returns true and you can see tls started in the connection object. Is this by design?:

    server = ldap3.Server(host=host, get_info='ALL', use_ssl=False)
    conn = ldap3.Connection(
        server=server,
        user=user_dn,
        password=password,
        raise_exceptions=True
    )

    print(conn)
    conn.open()
    print(conn)
    print(conn.start_tls())
    print(conn)
    conn.bind()
    print(conn)
    print(conn.extend.standard.who_am_i())

The output:

    not lazy - unbound - closed - <no socket> - tls not started - not listening - SyncStrategy - internal decoder
    not lazy - unbound - open - <local: *ip* - remote: *ip*:389> - tls not started - listening - SyncStrategy - internal decoder
    False
    not lazy - unbound - open - <local: *ip* - remote: *ip*:389> - tls not started - listening - SyncStrategy - internal decoder
    not lazy - bound - open - <local: *ip* - remote: *ip*:389> - tls not started - listening - SyncStrategy - internal decoder
    u:*ou*\*user*

**Comment by cannatag** [comment]
Hi, I'm not able to replicate this. I get True even with raise_exceptions set to True. Can you please give me additional information on how to replicate the issue? Thanks.

**Comment by ace6807** [comment]
I reinstalled ldap3 and started a new project just to rule out that being the problem. I'm still getting 'tls not started' and conn.start_tls() returning false when i set raise_exceptions=True in the Connection constructor. It does work properly if I explicitly set raise_exceptions=True after I call start_tls(). 

    import ldap3

    host = 'xxxxxxxxxxxxxxxxxxxx'
    user_dn = 'xxxx'
    user_pass = 'bad_password'

    server = ldap3.Server(host=host)
    conn = ldap3.Connection(
        server=server,
        user=user_dn,
        password=user_pass
    )

    print(conn)
    conn.open()
    print(conn)
    print(conn.start_tls())
    conn.raise_exceptions=True
    print(conn)
    conn.bind()
    print(conn)
    print(conn.bind())

Here you can see that it starts unbound with tls not started and tls starts fine. Then the exception gets raised for bad credentials as expected:

    not lazy - unbound - closed - <no socket> - tls not started - not listening - SyncStrategy - internal decoder
    not lazy - unbound - open - <local: ip:56744 - remote: ip:389> - tls not started - listening - SyncStrategy - internal decoder
    not lazy - unbound - open - <local: ip:56744 - remote: ip:389> - tls started - listening - SyncStrategy - internal decoder
    raise LDAPOperationResult(result=result['result'], description=result['description'], dn=result['dn'], message=result['message'], response_type=result['type'])
    ldap3.core.exceptions.LDAPInvalidCredentialsResult: LDAPInvalidCredentialsResult - 49 - invalidCredentials - None - 80090308: LdapErr: DSID-0C09042F, comment: AcceptSecurityContext error, data 52e, v2580  - bindResponse - None


Now if I move setting raise_exceptions=True to the constructor:

    import ldap3

    host = 'xxxxxxxxxxxxxxxxxxxx'
    user_dn = 'xxx'
    user_pass = 'bad_password'

    server = ldap3.Server(host=host)
    conn = ldap3.Connection(
        server=server,
        user=user_dn,
        password=user_pass,
        raise_exceptions=True
    )


    print(conn)
    conn.open()
    print(conn)
    print(conn.start_tls())
    print(conn)
    conn.bind()
    print(conn)
    print(conn.bind())

The expected error is still raised but tls never starts:

    not lazy - unbound - closed - <no socket> - tls not started - not listening - SyncStrategy - internal decoder
    not lazy - unbound - open - <local: ip:56770 - remote: ip:389> - tls not started - listening - SyncStrategy - internal decoder
    not lazy - unbound - open - <local: ip:56770 - remote: ip:389> - tls not started - listening - SyncStrategy - internal decoder
    raise LDAPOperationResult(result=result['result'], description=result['description'], dn=result['dn'], message=result['message'], response_type=result['type'])
    ldap3.core.exceptions.LDAPInvalidCredentialsResult: LDAPInvalidCredentialsResult - 49 - invalidCredentials - None - 80090308: LdapErr: DSID-0C09042F, comment: AcceptSecurityContext error, data 52e, v2580  - bindResponse - None

**Comment by cannatag** [comment]
If I run your latest code against the FreeIPA demo server I get:

```
ldap://ipa.demo1.freeipa.org:389 - cleartext - user: uid=admin,cn=users,cn=accounts,dc=demo1,dc=freeipa,dc=org - not lazy - unbound - closed - <no socket> - tls not started - not listening - SyncStrategy - internal decoder
ldap://ipa.demo1.freeipa.org:389 - cleartext - user: uid=admin,cn=users,cn=accounts,dc=demo1,dc=freeipa,dc=org - not lazy - unbound - open - <local: 192.168.1.100:63402 - remote: 52.57.162.88:389> - tls not started - listening - SyncStrategy - internal decoder
True
ldap://ipa.demo1.freeipa.org:389 - cleartext - user: uid=admin,cn=users,cn=accounts,dc=demo1,dc=freeipa,dc=org - not lazy - unbound - open - <local: 192.168.1.100:63402 - remote: 52.57.162.88:389> - tls started - listening - SyncStrategy - internal decoder
Traceback (most recent call last):
  File "C:/Users/Giovanni Cannata/.PyCharm2017.3/config/scratches/scratch_17.py", line 21, in <module>
    conn.bind()
  File "C:\Python\OpenSource\ldap3\ldap3\core\connection.py", line 530, in bind
    response = self.post_send_single_response(self.send('bindRequest', request, controls))
  File "C:\Python\OpenSource\ldap3\ldap3\strategy\sync.py", line 124, in post_send_single_response
    responses, result = self.get_response(message_id)
  File "C:\Python\OpenSource\ldap3\ldap3\strategy\base.py", line 381, in get_response
    raise LDAPOperationResult(result=result['result'], description=result['description'], dn=result['dn'], message=result['message'], response_type=result['type'])
ldap3.core.exceptions.LDAPInvalidCredentialsResult: LDAPInvalidCredentialsResult - 49 - invalidCredentials - None - None - bindResponse - None

Process finished with exit code 1
```

The start_tls() operation is executed and returns True. I don't see this in your output, can you try your code with the following environment please:

```
host = 'ipa.demo1.freeipa.org'
user_dn = 'uid=admin,cn=users,cn=accounts,dc=demo1,dc=freeipa,dc=org'
user_pass = 'bad_password'
```

You should get the ``True`` answer from the start_tls() operation.

**Comment by cannatag** [comment]
Hi, the problem is only with AD, that seems to not allow the reading of the schema with an anonymous bind (that is performed with open() ). This is silently passing while the raise_exceptions is False, but was caught by  raise_exceptions=True but the exception was not really raised. Now the exception is not raised even with raise_exceptions=True because the server can decide to not send the schema anyway.

Please check code in dev to see if it's working as expected.

Thanks,
Giovanni

**Comment by pyther** [comment]
I'll add that I'm also experiencing this issue with an openldap server.

**Comment by cannatag** [comment]
Released in version 2.4.1
